extensions-core/multi-stage-query/src/main/java/org/apache/druid/msq/querykit/BaseLeafFrameProcessorFactory.java

```diff
@@ -146,32 +142,91 @@ public ProcessorsAndChannels<Object, Long> makeProcessors(
+          CollectionUtils.getOnlyElement(segmentMapFnList, throwable -> DruidException.defensive("Only one segment map function expected"));
+      return createBaseLeafProcessorManagerWithHandoff(
+          stageDefinition,
+          inputSlices,
+          inputSliceReader,
+          counters,
+          warningPublisher,
+          segmentMapFunction,
+          frameWriterFactoryQueue,
+          channelQueue,
+          frameContext
+      );
+    };
 
     //noinspection rawtypes
     final ProcessorManager processorManager;
 
     if (segmentMapFnProcessor == null) {
       final Function<SegmentReference, SegmentReference> segmentMapFn =
           query.getDataSource().createSegmentMapFunction(query, new AtomicLong());
-      processorManager = processorManagerFn.apply(segmentMapFn);
+      processorManager = processorManagerFn.apply(ImmutableList.of(segmentMapFn));
     } else {
-      processorManager = new ChainedProcessorManager<>(segmentMapFnProcessor, processorManagerFn);
+      processorManager = new ChainedProcessorManager<>(ProcessorManagers.of(() -> segmentMapFnProcessor), processorManagerFn);
     }
 
     //noinspection unchecked,rawtypes
     return new ProcessorsAndChannels<>(processorManager, OutputChannels.wrapReadOnly(outputChannels));
   }
 
+  private ProcessorManager<Object, Long> createBaseLeafProcessorManagerWithHandoff(
+      final StageDefinition stageDefinition,
+      final List<InputSlice> inputSlices,
+      final InputSliceReader inputSliceReader,
+      final CounterTracker counters,
+      final Consumer<Throwable> warningPublisher,
+      final Function<SegmentReference, SegmentReference> segmentMapFunction,
+      final Queue<FrameWriterFactory> frameWriterFactoryQueue,
+      final Queue<WritableFrameChannel> channelQueue,
+      final FrameContext frameContext
+  )
+  {
+    final BaseLeafFrameProcessorFactory factory = this;
+    // Read all base inputs in separate processors, one per processor.
+    final Iterable<ReadableInput> processorBaseInputs = readBaseInputs(
+        stageDefinition,
+        inputSlices,
+        inputSliceReader,
+        counters,
+        warningPublisher
+    );
+
+    return new ChainedProcessorManager<>(
+        new BaseLeafFrameProcessorManager(
+            processorBaseInputs,
+            segmentMapFunction,
+            frameWriterFactoryQueue,
+            channelQueue,
+            frameContext,
+            factory
+        ),
+        objects -> {
+          if (objects.isEmpty()) {
+            return ProcessorManagers.none();
+          }
+          List<InputSlice> handedOffSegments = new ArrayList<>();
+          for (Object o : objects) {
+            if (o instanceof SegmentsInputSlice) {
+              SegmentsInputSlice slice = (SegmentsInputSlice) o;
+              handedOffSegments.add(slice);
+            }
+          }
+          return new BaseLeafFrameProcessorManager(
```

Comment:
I think we are fetching stuff from deep storage here. Lets add a comment for future maintenance.